crates/worker/src/partition_processor_manager.rs

```diff
@@ -684,68 +763,80 @@ impl<T: TransportConnect> PartitionProcessorManager<T> {
+        events: EventSender,
     ) -> Result<(), Error> {
         let partition_id = control_processor.partition_id;
 
         match control_processor.command {
             ProcessorCommand::Stop => {
-                if let Some(processor) = self.running_partition_processors.remove(&partition_id) {
-                    if let Some(handle) = self.task_center.cancel_task(processor.task_id) {
-                        debug!(%partition_id, "Asked by cluster-controller to stop partition");
-                        if let Err(err) = handle.await {
-                            warn!("Partition processor crashed while shutting down: {err}");
-                        }
-                    }
-                } else {
-                    debug!("No running partition processor. Ignoring stop command.");
+                if let Some(status) = self.running_partition_processors.remove(&partition_id) {
+                    status.stop(&self.task_center).await;
                 }
             }
-            ProcessorCommand::Follower => {
-                if let Some(state) = self.running_partition_processors.get_mut(&partition_id) {
-                    // if we error here, then the system is shutting down
-                    state.step_down()?;
-                } else if let Some(partition_key_range) = partition_table
-                    .get_partition(&partition_id)
-                    .map(|partition| &partition.key_range)
-                {
-                    self.start_partition_processor(
-                        partition_id,
-                        partition_key_range,
-                        RunMode::Follower,
-                    )
-                    .await?;
+            ProcessorCommand::Follower | ProcessorCommand::Leader => {
+                let run_mode = if control_processor.command == ProcessorCommand::Leader {
+                    RunMode::Leader
+                } else if control_processor.command == ProcessorCommand::Follower {
+                    RunMode::Follower
                 } else {
-                    debug!("Unknown partition id '{partition_id}'. Ignoring follower command.");
-                }
-            }
-            ProcessorCommand::Leader => {
-                if let Some(state) = self.running_partition_processors.get_mut(&partition_id) {
-                    state
-                        .run_for_leader(
-                            self.metadata_store_client.clone(),
-                            self.metadata.my_node_id(),
-                        )
-                        .await?;
+                    unreachable!()
+                };
+
+                if let Some(status) = self.running_partition_processors.get_mut(&partition_id) {
+                    if let ProcessorStatus::Started(state) = status {
+                        // if we error here, then the system is shutting down
+                        if run_mode == RunMode::Follower {
+                            state.step_down()?;
```

Comment:
Just wanna verify something here that came to my mind when reading the ingress <-> PP code.

With this PR do you make asynchronous the "stopping"/"become follower" process too? If that's the case, can we now incur in a situation where network messages are in-flight to the PP **while** the PP is stopping/becoming follower, thus ending up with those messages essentially stuck between channels?

Reply:
If I get your concerns right, then the answer is No. stepping_down or running_for_leader both runs on the main event loop, thus no rpc messages are processed. Only starting new PP runs asynchronously 

That being said, both stepping down and running for leader happens by pushing a message to the PP handler channel. But that has been always the case.

Reply:
https://github.com/restatedev/restate/pull/2254